src/mlpack/methods/kde/kde_impl.hpp

```diff
@@ -418,6 +456,16 @@ Evaluate(Tree* queryTree,
                                 "dual-tree");
   }
 
+  // Clean accumulated alpha if Monte Carlo estimations are available.
+  if (monteCarlo && std::is_same<KernelType, kernel::GaussianKernel>::value)
+  {
+    Timer::Start("cleaning_query_tree");
+    KDECleanRules<Tree> cleanRules;
+    DualTreeTraversalType<KDECleanRules<Tree>> cleanTraverser(cleanRules);
```

Comment:
I'd suggest using two single-tree traversals here; the reason is that a dual-tree traversal will compare every pair of nodes in the query tree and reference tree (`O(N^2)` where each tree has `O(N)` nodes), but a single-tree traversal will fix one tree in `O(N)` time.  So the scaling should be better with two individual single-tree recursions.

Reply:
You're right! And actually we only need to clean the query tree :)

Reply:
Ah, good point. :+1: